app/models/product.rb

```diff
@@ -1,20 +1,47 @@
 class Product < ApplicationRecord
+  PERMALINK_REGEX = /[[:alnum:]]+/
+  validates_with ImageUrlValidator, attributes: [:image_url], if: ->{ image_url.present? }
+
   has_many :line_items
   has_many :orders, through: :line_items
   before_destroy :ensure_not_referenced_by_any_line_item
 
-  validates :title, :description, :image_url, presence: true
-  validates :price, numericality: { greater_than_or_equal_to: 0.01 }
+  validates :title, :description, :image_url, :price, presence: true
+
+  validates :price, numericality: { greater_than_or_equal_to: 0.01 }, if: :price_present?
+  validates_comparison_of :price, greater_than: :discount_price, message: "should be greater than discount price", if: [:price_present?, :discount_price_present?]
   validates :title, uniqueness: true
-  validates :image_url, allow_blank: true, format: {
-    with: %r{\.(gif|jpg|png)\z}i,
-    message: 'must be a URL for GIF, JPG or PNG image.'
-  }
-
-  private def ensure_not_referenced_by_any_line_item
-    unless line_items.empty?
-      errors.add(:base, 'Line Items present')
-      throw :abort
+
+  validates :permalink, uniqueness: true, format: { with: PERMALINK_REGEX, message: "no special and no space allowed in the permalink" }, if: :permalink_present?
```

Comment:
let update the message to `no special character or space .....(same as already written)`

Reply:
modified message content and used permalink? instead of permalink_present?